Components: Create Circular Progress Bar for Launchpad

sharpirate opened the pull request:
#### Proposed Changes
Create a new `circular-progress-bar` component and display it on the Launchpad screen

<img width="433" alt="image" src="https://user-images.githubusercontent.com/20927667/208856066-9462d91a-2a5f-4795-93d2-d0af59702687.png">

#### Testing Instructions
You can test this with any Launchpad flow.

1. Checkout this branch
2. Go to `http://calypso.localhost:3000/setup/link-in-bio/intro`
3. Complete the flow until you arrive at Launchpad
4. You should see the new circular progress bar showing the number of completed tasks and the number of total tasks
5. Complete a task
6. You should see both the numbers and the progress circle on the progress bar update accordingly

https://user-images.githubusercontent.com/20927667/208854642-03151d2d-5bd0-41c1-a584-b460ecd3e6b4.mov


<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #71326

simison commented:
Any possibility to add this to `@automattic/components` package if this isn't too specific to launchpad?

sharpirate committed:
add progressbar role

sharpirate committed:
move component to @automattic/components

Sharpirate commented:
@simison Thanks, I was thinking I should do that. I just updated the PR to move the component to `@automattic/components`

Sharpirate commented:
### Notes
The current implementation is not 100% matching the design. Here are the differences:
1. In the design the font weight is 500 but it doesn't look like we have all of the different font weights available to us right now. I tried using the 600 font weight but it looks too thick so for now I am sticking with the 100. Let me know if that looks good.

![font-weights](https://user-images.githubusercontent.com/20927667/208858134-a21cdc58-19b8-41cb-8c47-608b0d1a0f62.png)

2. In the design the stroke width for the gray circle and the blue progress circle are different: 3px and 5px respectively. However, this is really tricky to implement in the code because that means both circles should have different radiuses and we start getting issues with the svg being cropped and scaled unevenly. For now, I kept both circles at 5px. Let me know if that looks good.

3. In the design we have a few different sizes for the progress bar. We have `40px`, `48px` and `52px`. Are we planning to have these three different versions so we can choose the appropriate size depending on where we render the progress bar (Launchpad, Site-Editor e.g.) or is the idea that the size should adjust automatically on smaller screens? Because if it's the latter this will be a problem too, because it's not as simple as changing CSS rules based on screen size. We are actually using the size in the code to calculate the length of the arc of the progress circle. So we need to know the size of the element before hand. For now, I have set the size to a static `48px`. Let me know if that looks good.

@nuriapenya Curious what your thoughts are on all of these

simison commented:
> or is the idea that the size should adjust automatically on smaller screens? 

Does `transform: scale(20)` solve above problem? You might still want to optimize pixel sizes and stick to specific sizes, though.

nuriapenya commented:
Thanks, @Sharpirate for building this so fast! 💎 

>I am sticking with the 100

Sounds good. Initially, I was using `Medium`, which I think is the equivalent of 300-400. But if that's not possible, I'd tend to go on thin rather than thick.

> For now, I kept both circles at 5px. 

Also fine. The main objective was to create a nice distinction between the two levels, but we can ditch it for now if this is too complex.

>or is the idea that the size should adjust automatically on smaller screens?

That would be my preference. Especially since we plan to use this element in different areas, and they might require a bit of scaling up or down. But we can have it fixed for now. Could we do at least two sizes? One for desktop and another for mobile?

Desktop: 48px (font-size: 15px)
Mobile: 40px (font-size: 12px)

nuriapenya commented:
@Sharpirate One small correction from my side: 

Let's make the stroke of all circles 4px. This way looks more balanced both on desktop & mobile. I updated [all the designs](9wpoYJrMlYKAKUYvVBMN74-fi-1760%3A43986).

<img width="1145" alt="Screenshot 2022-12-21 at 12 35 27" src="https://user-images.githubusercontent.com/7931304/208896469-e8501d1e-9856-481e-af56-2921af8a1f8a.png">

Sharpirate commented:
> Does transform: scale(#) solve above problem?

@simison Thanks for the suggestion, that will definitely work!

sharpirate committed:
add mobile size & change stroke size

Sharpirate commented:
Thanks, @nuriapenya! I've updated the PR to have the stroke width at `4px` and added a mobile size of `40px` (and desktop at `48px`):

![Circular Progress Bar Mobile Size](https://user-images.githubusercontent.com/20927667/209008007-205098db-b463-4c86-a573-844a0817c55e.gif)

Actually with Mikael's suggestion of using CSS `transform: scale` we can have different sizes easily. Let me know if we should add more sizes and for what screen sizes or if we should stick with `40px` mobile and `48px` desktop for now

sharpirate committed:
rebase with trunk

sharpirate committed:
update README

nuriapenya commented:
>Actually with Mikael's suggestion of using CSS transform: scale we can have different sizes easily. Let me know if we should add more sizes and for what screen sizes or if we should stick with 40px mobile and 48px desktop for now

Looking cool! I'd say, for now, let's leave the two fixed sizes. Once (and if) we start to use this element in more often, we can always add the `transform: scale` property.

The desktop font should be smaller. Can we try with 13px? 🙏🏽  

Thanks a lot @Sharpirate

edanzer commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/sidebar.tsx:
```diff
@@ -37,16 +37,19 @@ function getUrlInfo( url: string ) {
 	}, 0 );
 
-	return Math.round( ( totalCompletedTasks / tasks.length ) * 100 );
+	const totalTasks = tasks.length;
```
Worth noting that calculating and returning totalTasks here might be a bit redundant. You could already get that directly from from enhancedTasks.length without this method. That said, it's also tidy doing it this way, so works for me.

sharpirate replied:
@edanzer That is a great suggestion and simplifies the function a lot (returning a single value instead of an array)

edanzer commented:
This tests well for me! 
- [x] Icon looks like it should based on screenshots/gifs above.
- [x] Responsive behavior matches gif above.
- [x] Code changes look good, too.
- [x] Nice to see the documentation, and the documentation is good.
- [x] I also checked our launchpad tests. We have one for `displays a progress bar based off of task completion` that I thought might fail now, but looks like that also still passes since you've added the role=progressBar to the new icon. 

Approving with a these notes: 
   - I think we still need to make desktop font changes requested by @nuriapenya above. 
   - There are some failing tests. And given amount of time since this PR was initiated, it might be worth rebasing. If you do that and want another test, let me know.

sharpirate committed:
change font size & simplify getTasksProgress func

Sharpirate commented:
@nuriapenya 
>The desktop font should be smaller. Can we try with 13px? 
Looks like the stylelint is forcing me to use `rem` units for the font size and the closest it's allowing me to do is `0.75rem` which is `12px`, would that work?

Addison-Stavlo commented on packages/components/src/circular-progress-bar/README.md:
```diff
@@ -0,0 +1,22 @@
+# Circular Progress Bar
+
+This component is used to display a circular progress bar as part of the cohesive Launchpad Experience
+See https://cylonp2.wordpress.com/2022/12/14/a-cohesive-launchpad-experience/
+It displays a gray circle and another blue circle on top of it showing the progress.
+It renders text in the center in the center: X/Y where X is the current step and Y is the total number of steps
+The currentStep and numberOfSteps props are used to calculate represent the current progress as a fraction
+
```
I wonder if we should name this something like "LaunchPad Circular Progress Bar", or put a big warning that this is intended for launchpad checklists (in their various places) and not necessarily for other unrelated progress bar solutions. I say that since our launchpad/checklist icons used for distinguishing and returning will all use the similar distinguishable circular progress icon, and if other interfaces start using the circular progress bar format as well that could get confusing.

I see you already noted "This component is used to display a circular progress bar as part of the cohesive Launchpad Experience", but maybe good to add that its not intended for non-launchpad related interfaces.

Addison-Stavlo commented:
This looks great and works well in testing.

Other than whats noted above, I just wanted to say I think this is a really cool progress bar. Kudos to both Nuria and Bobby here!

sharpirate committed:
update component README

sharpirate committed:
update readme
